ffi/capi/dart/package/lib/src/ICU4XDateTimeFormatter.dart

```diff
@@ -0,0 +1,137 @@
+        ICU4XDateLengthAsFfi(dateLength),
+        ICU4XTimeLengthAsFfi(timeLength));
+    return result.isOk
+        ? ICU4XDateTimeFormatterFromFfi(result.union.ok)
+        : throw ICU4XErrorFromFfi(result.union.err);
+  }
+  static late final _createWithLengthsFfi = capi<
+          ffi.NativeFunction<
+              ResultICU4XDateTimeFormatterUnionICU4XError Function(
+                  ICU4XDataProviderFfi,
+                  ICU4XLocaleFfi,
+                  ffi.Uint32,
+                  ffi.Uint32)>>('ICU4XDateTimeFormatter_create_with_lengths')
+      .asFunction<
+          ResultICU4XDateTimeFormatterUnionICU4XError Function(
+              ICU4XDataProviderFfi,
+              ICU4XLocaleFfi,
+              ICU4XDateLengthFfi,
+              ICU4XTimeLengthFfi)>();
+
+  /// Formats a [`ICU4XDateTime`] to a string.
+  ///
+  /// See the [Rust documentation for `format`](https://docs.rs/icu/latest/icu/datetime/struct.DateTimeFormatter.html#method.format) for more information.
+  String formatDatetime(ICU4XDateTime value) {
+    final writeable = create_writeable();
+    final result = _formatDatetimeFfi(
+        this._underlying, ICU4XDateTimeAsFfi(value), writeable);
+    return result.isOk
+        ? writeable_to_string(writeable)
+        : throw ICU4XErrorFromFfi(result.union.err);
+  }
+
+  static late final _formatDatetimeFfi = capi<
+          ffi.NativeFunction<
+              ResultVoidUnionICU4XError Function(
+                  ICU4XDateTimeFormatterFfi,
+                  ICU4XDateTimeFfi,
+                  DiplomatWriteable)>>('ICU4XDateTimeFormatter_format_datetime')
+      .asFunction<
+          ResultVoidUnionICU4XError Function(ICU4XDateTimeFormatterFfi,
+              ICU4XDateTimeFfi, DiplomatWriteable)>();
+
+  /// Formats a [`ICU4XIsoDateTime`] to a string.
+  ///
+  /// Will convert to this formatter's calendar first
+  ///
+  /// See the [Rust documentation for `format`](https://docs.rs/icu/latest/icu/datetime/struct.DateTimeFormatter.html#method.format) for more information.
+  String formatIsoDatetime(ICU4XIsoDateTime value) {
+    final writeable = create_writeable();
+    final result = _formatIsoDatetimeFfi(
+        this._underlying, ICU4XIsoDateTimeAsFfi(value), writeable);
+    return result.isOk
+        ? writeable_to_string(writeable)
+        : throw ICU4XErrorFromFfi(result.union.err);
+  }
+
+  static late final _formatIsoDatetimeFfi = capi<
+              ffi.NativeFunction<
+                  ResultVoidUnionICU4XError Function(ICU4XDateTimeFormatterFfi,
+                      ICU4XIsoDateTimeFfi, DiplomatWriteable)>>(
+          'ICU4XDateTimeFormatter_format_iso_datetime')
+      .asFunction<
+          ResultVoidUnionICU4XError Function(ICU4XDateTimeFormatterFfi,
+              ICU4XIsoDateTimeFfi, DiplomatWriteable)>();
+
+  ICU4XDateTimeFormatter._(this._underlying) {
+    _finalizer.attach(this, this._underlying.cast());
+  }
+
+  static late final _finalizer = ffi.NativeFinalizer(
```

Comment:
Question: when exactly does Dart execute the finalizer? When the object is GC'd, or when the object leaves scope? Does this differ on native and web?

Reply:
It's at GC time, but `NativeFinalizer` at least guarantees that it happens.

Reply:
https://api.flutter.dev/flutter/dart-ffi/NativeFinalizer-class.html#:~:text=Compared%20to%20the,recognized%20as%20inaccessible.